### PR TITLE
Add spelling checks to content

### PR DESCRIPTION
**Pull request by brianamarie**
## Overview
**TL;DR**
Our tests don't currently check for spelling or other grammar like things. And they can! And they should! We have the power! 💪 

### Questions
- Does anyone have experience/knowledge of good libraries or ways to start? 
- It appears we will have to use npm alongside gems. Can we have a project.json file? Do we need one? 

### Next Steps
- ~~Remove writewell test, because it's erroring~~
- [x] Mimic testing in successful repository
- [x] Figure out how to call writewell succesfully in `script/cibuild` to check all `.html` and `.md` files

### Review
Not ready for review yet, but I very much welcome help! 

cc/  @github/services-training and @github/services-engineering for help or expertise if willing 🙇

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,5 +1,13 @@
 language: ruby
 rvm:
   - 2.0
+  
+install:
+- . $HOME/.nvm/nvm.sh
+- nvm install stable
+- nvm use stable
+- npm install writewell -g
```
Comment (stoe) [inline .travis.yml]: `npm install -g writewell`

**Comment by stoe** [comment]
> It appears we will have to use npm alongside gems. Can we have a project.json file? Do we need one?

@brianamarie, yes, we should have one.

```json
{
  "name": "training-kit",
  "version": "1.0.0",
  "description": "Open source on demand courses and cheat sheets for Git and GitHub",
  "keywords": [
    "git",
    "github",
    "cheat-cheets",
    "training"
  ],
  "author": "GitHub Teacher <training@github.com>",
  "contributors": [
    "Brina Swift <brianamarie@github.com>",
    "Stefan Stölzle <stefan@github.com>"
  ],
  "license": "CC-BY-4.0",
  "homepage": "https://services.github.com/on-demand/",
  "repository": "https://github.com/github/training-kit.git",
  "bugs": {
    "url": "https://github.com/github/training-kit/issues"
  },
  "scripts": {
    "test": "writewell -l script/list-of-files.txt"
  },
  "dependencies": {
    "writewell": "^0.7.1"
  }
}
``` 

And then `.travis.yml`:

```yml
language: ruby
rvm:
  - 2.0
  
install:
  - . $HOME/.nvm/nvm.sh
  - nvm install stable
  - nvm use stable
  - npm install

script:
  - script/cibuild
  - npm test
```

**Comment by stoe** [comment]
Just looked it up, and found that the `npm install -g` should be in `before_script`:

```yml
install:
- . $HOME/.nvm/nvm.sh
- nvm install stable
- nvm use stable

before_script:
  - npm install writewell -g

script:
```

**Comment by brianamarie** [comment]
Test currently works, but there are a few things to sort: 

- Set up configs to check only for what we want to check for
- Figure out how to check all HTML and md files

**Comment by brianamarie** [comment]
:tada: Alright, a custom dictionary has been created in `assets/dic/extra-words.dic`. The number at the top represents how many words exist in the file, so please update that number as you add or delete files. Here is some [good documentation](http://manpages.ubuntu.com/manpages/trusty/man4/hunspell.4.html) on how to interact with this dictionary file. I will continue to work with that file and the settings to adjust them to actually detect things we want them to. 👍

File: settings.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
+{
+  "dictionaries": {
+    "build-in": ["en_US", "en_GB"],
+    "custom": ["assets/dic/extra-words.dic"]
```
Comment (stoe) [inline settings.json]: > a custom dictionary has been created in `assets/dic/extra-words.dic`

@brianamarie, I think we should spell out folder names, `assets/dictionary/extra-words.dic`

Comment (brianamarie) [inline settings.json]: Great point, @stoe. Addressed in [4ea1a47](https://github.com/github/training-kit/pull/446/commits/4ea1a473d6fdd867efaa012319dfb2f0c47f79db).

File: script/list-of-files.txt [hunk 3]
```diff
@@ -1,3 +1,4 @@
 ./README.md
 ./_site/CONTRIBUTING.md
 ./_site/index.html
+find . -type f -name "*.html"
```
Comment (stoe) [inline script/list-of-files.txt]:
```sh
find . -name '*.html'
```

@brianamarie, should be enough, it's always a file with an extension.

> Find files by extension:
> `find {{root_path}} -name '{{*.ext}}'`

via https://github.com/tldr-pages/tldr/blob/master/pages/common/find.md

---

https://github.com/tldr-pages/tldr is my favourite CLI tools to quickly look up these things

File: .travis.yml [hunk 4]
```diff
@@ -1,5 +1,20 @@
...
+install:
+  - rm -rf ~/.nvm && git clone https://github.com/creationix/nvm.git ~/.nvm && (cd ~/.nvm && git checkout `git describe --abbrev=0 --tags`) && source ~/.nvm/nvm.sh && nvm install $TRAVIS_NODE_VERSION
+  - npm install -g writewell
+
 script:
   - script/cibuild
+  - cd /_site
+  - find . -type f -name "*.html" >> ./script/list-of-files.txt
+  - cd ..
+  - npm test
```
Comment (stoe) [inline .travis.yml]:
```yml
  - script/cibuild
  - cd /_site
  - find . -type f -name "*.html" >> ./script/list-of-files.txt
  - cd ..
- npm test
```

The last 3 steps should be part of the `script/cibuild` file.

Comment (brianamarie) [inline .travis.yml]: @stoe I think one of us read the other's mind, I just committed that in the most recent commit! 👍 😄

File: script/cibuild [hunk 5]
```diff
@@ -10,7 +10,13 @@ ruby -v
...
 echo "pwd: $(pwd)"
 git --version
+echo "node: $(node --version)"
+echo "npm: $(npm --version)"
 
 script/bootstrap
 
 bundle exec rake test
+
+find ./_site -type d \( -name assets -o -name _javascript -o -wholename downloads/es_ES/github-git-cheat-sheet/index.html \) -prune -o -name "*.html" -print >> ./script/list-of-files.txt
```
Comment (brianamarie) [inline script/cibuild]: The first two parts of this `-prune` work fine, omitting files from the assets file and the _javascript file. BUT, I would really like to include specific cheat sheets, as well. Any suggestions on how to do this?

Comment (swinton) [inline script/cibuild]: @brianamarie probably there are multiple solutions... One way would be to manually add them to the end of the  `list-of-files.txt`, after the `find` operation e.g.

```bash
echo foo >> list-of-files.txt
echo bar >> list-of-files.txt
```

**Comment by stoe** [comment]
> The first two parts of this -prune work fine, omitting files from the assets file and the _javascript file. BUT, I would really like to include specific cheat sheets, as well. Any suggestions on how to do this?

@brianamarie, try ...

```sh
find ./_site -type d \( -path ./_site/_javascript -o -path ./_site/assets -o -path ./_site/downloads/es_ES \) -prune -o -name "*.html" -print
``` 

"Just" add all cheatsheet languages like ` -o -path ./_site/downloads/es_ES`

**Comment by brianamarie** [comment]
@stoe I believe I have a workaround with this:

```
find ./_site -type d \( -name assets -o -name _javascript -o -name downloads \) -prune -o -name "*.html" -print >> ./script/list-of-files.txt
find ./_site/downloads -type d \( -name ar -o -name bn_BD -o -name de -o -name es_ES -o -name fr -o -name it -o -name ja -o -name kr -o -name pt_BR -o -name pt_PT -o -name sk -o -name zh_CN \) -prune -o -name "*.html" -print >> ./script/list-of-files.txt
```

But, if there is a benefit to condensing it to one line, I'm all for it! 👍

**Comment by stoe** [comment]
@brianamarie, just saw you added them all in b0c365e.

Not :100: if we should use `-name` as it could create false positives.
I'd rather use `-path`, but it's your call.

**Comment by brianamarie** [comment]
> I'd rather use -path, but it's your call.

@stoe Ok. I'm not super familiar with -name vs -path. I will look into this. I gather I can just replace -name with -path and the core functionality will be the same?

**Comment by stoe** [comment]
> I gather I can just replace -name with -path and the core functionality will be the same?

I think it has to be a relative path from where the script runs, e.g `-path ./_script/downloads/es_ES`.

All in all, the line would be longer. That's why I think we should refactor the entire thing into its own `script/xxx` file.
If you want I can look into that tomorrow. That way we might also be able to keep an easy to change "white-/blacklist" of what to include/exclude.

**Comment by brianamarie** [comment]
> If you want I can look into that tomorrow. That way we might also be able to keep an easy to change "white-/blacklist" of what to include/exclude.

@stoe If you have time, that'd be fantastic! I like the idea of having an easy to change "white-/blacklist". 💯

**Comment by stoe** [comment]
Just noticed another thing with having it in two lines.

The first line already adds what we're trying to exclude in the second. 😒

**Comment by brianamarie** [comment]
> The first line already adds what we're trying to exclude in the second. 😒

It shouldn't...it should be omitting the downloads file completely, I think?

**Comment by brianamarie** [comment]
Next steps:

- Figure out why it's saying the test is failing, with no explanation 
- Decide as a team ( @github/services-training ) what 'write-good' filters we want. Right now, they're all off. (Set to false in the `settings.json` file.) 
- @stoe's improvement to process for omitting certain files from these tests

**Comment by brianamarie** [comment]
@stoe What do you think about merging in its current state?

**Comment by stoe** [comment]
@brianamarie, sounds good. 
Go for it and :shipit:

**Comment by hectorsector** [comment]
@stoe I did some testing on `[test-write-good](https://github.com/github/training-kit/commits/test-write-good)` and the tests are passing even though there's some spelling mistakes. When I run the write-well script locally, there's no output from the script.

**Comment by stoe** [comment]
@hectorsector, sorry, but I think I can't help here.

Maybe @brianamarie knows as she's the one who started this PR.

**Comment by brianamarie** [comment]
@stoe It looks like the tests stopped running in commit [8ff52](https://github.com/github/training-kit/pull/446/commits/8ff525fa08694e12f0a51f92830694973aa831b1). Does this make sense to you?

**Comment by stoe** [comment]
It's breaking again, woohoo.
Over to you, @brianamarie.

**Comment by brianamarie** [comment]
@hectorsector Do you have any thoughts on why this build is failing? It seems to be something wrong with writewell, but I can't nail down exactly what it is. Or, is there anyone in @github/services-engineering that wouldn't mind taking a peek at these tests?

**Comment by stoe** [comment]
@brianamarie, there are writwell errors reported: https://travis-ci.org/github/training-kit/builds/201529765#L1342-L1360

File: advanced/index.md [hunk 6]
```diff
@@ -1166,7 +1169,6 @@ A sample `.gitconfig` file:
 This course covers many advanced uses of Git and GitHub, and yet there is still more to explore. We've included some of the most useful resources for our students with insatiable appetites.
 
 #### Advanced Git Videos
-* [Advanced Git, presented at JavaZone](http://vimeo.com/49444883)
```
Comment (stoe) [inline advanced/index.md]: Hm... https://vimeo.com/49444883 exists, looks like a false positive.

Comment (stoe) [inline advanced/index.md]: Think it just has to be `https://` to work ¯\_(ツ)_/¯:

**Comment by brianamarie** [comment]
@stoe, good point that it's a false positive. I tried adding back in with `https` and got the same problem. Since this page is deprecated in general and not linked to, I'm going to remove that link for now.

File: Rakefile [hunk 7]
```diff
@@ -1,11 +1,39 @@
...
+      :url_ignore  => [
+        %r{https?://github.com/new},
+        /\.github\.io/,
+        /\.gitter\.im/,
+        /atom\.io/,
+        /git-scm\.com/,
+        /github\.com/,
+        /mademistakes\.com/,
+        /nasa\.gov/,
+        /travis-ci\.org/,
```
Comment (stoe) [inline Rakefile]: @brianamarie, if you think it's necessary, we can add `/vimeo\.com/` here as an exception for the link checker.

**Comment by brianamarie** [comment]
@kylemacey (or any @github/services-engineering) Do you happen to know anything about these errors:

`Errno::EMFILE: Too many open files - getcwd`  (locally)
`*** Error in node': malloc(): memory corruption: 0x0000000002a8bdf0 ***
Aborted (core dumped)` (on Travis)

**Comment by brianamarie** [comment]
I'm trying to investigate the errors, and I'm wondering if they have to do with node-gyp. Some of what I've read online says that having an updated Node version can fix this. I'm trying out these things on a separate branch, `brianamarie/temp-branch-DONOTMERGE-write-good`, to keep this PR less noisy. I'll post here if I figure anything out!

**Comment by migarjo** [comment]
Hi @brianamarie, I tried to take a look at it but wasn't able to see anything. @stoe, I know you looked at this earlier, but it looks like things have progressed since then. Are you familiar with this problem?

**Comment by stoe** [comment]
@migarjo @brianamarie all I can find googling the error is that it looks like a Ruby error, because there are too many open files in the process.

That it's Ruby might be because we run our node script within Ruby.

**Comment by brianamarie** [comment]
Hi @latiranitaardiente! Check out the open issues, and our README.md to see if there are any ways you want to help contribute! 😄

**Comment by brianamarie** [comment]
Currently very confused why _this_ branch is getting weird memory corruptions, when [this branch](https://travis-ci.org/github/training-kit/builds/202422136) doesn't have that many differences, but isn't. 🤔

File: script/cibuild [hunk 8]
```diff
@@ -2,15 +2,50 @@
...
+export RUBY_GC_HEAP_INIT_SLOTS=800000
+export RUBY_HEAP_FREE_MIN=100000
+export RUBY_HEAP_SLOTS_INCREMENT=400000
+export RUBY_HEAP_SLOTS_GROWTH_FACTOR=1
+
 export PATH="/usr/share/rbenv/shims:$PATH"
+export RACK_ROOT=$(cd "$(dirname $0)"/.. && pwd)
+export RACK_ENV="test"
+export RAILS_ENV="test"
+export RBENV_VERSION="2.2.3"
```
Comment (hectorsector) [inline script/cibuild]: @brianamarie @stoe is there a particular reason we're testing against this version? I get the following when I try to test with this version: 

```sh
hostname: Hector's
pwd: /Users/halfaro1/github/training-kit
git version 2.11.0

==> Installing gem dependencies…
rbenv: bundle: command not found

The `bundle' command exists in these Ruby versions:
  2.0.0-github6
  2.3.0
  2.3.1

rbenv: bundle: command not found

The `bundle' command exists in these Ruby versions:
  2.0.0-github6
  2.3.0
  2.3.1
```

**Comment by zalew16** [comment]
+ -. $ HOME / .nvm / nvm.sh

+ - NVM zainstalować stabilna

+ - Wykorzystanie NVM stabilna

+ - Npm zainstalować writewell -g

**Comment by brianamarie** [comment]
Closing - it was a side branch for testing on an effort that is put in the backlog for the time being.
